### DISCUSSION TITLE
Executing `retry` bound to `RetryHandler` instance instead of `retryOpts`

### DISCUSSION THREAD

**Discussion by mikaelkaron** [nodejs/undici#2973] [Ideas]
At the moment the `retry` handler is [executed bound to `retryOpts`](https://github.com/nodejs/undici/blob/main/lib/handler/retry-handler.js#L313C20). This prevents access to instance variables and methods on the `RetryHandler` instance, and I'm wondering if simply doing:

```javascript
    this.retryOpts.retry.call(
      this,
      err,
      {
        state: { counter: this.retryCount++, currentTimeout: this.retryAfter },
        opts: { retryOptions: this.retryOpts, ...this.opts }
      },
      onRetry.bind(this)
    )
 ```

Would be an acceptable solution to this.

**Comment by metcoder95** [comment]
👋 
That's the intention, to hide internals of the `RetryHandler` itself; but curious what's the use case you are looking to enable?

**Reply by mikaelkaron** [reply]
Om writing an interceptor to do digest auth and wanted the headers from the `retry` and a simple way to pass them to `dispatch`. I solved it with a closure, so don’t have an active use-case right now but figured I’d check in to see if the bound context was intentional. 

Is there another recommended way to pass data from the `retry` method other than a custom closure? The only problem I have with it is that it prevents me from reusing the `RetryHandler` across requests.

**Reply by metcoder95** [reply]
The `RetryHandler` is designed to be bound to a single request, as it is not made as a reliability mechanism but rather a flow control one in the case of a bad response.
Though, is maybe something worth it to explore as I've seen few request for them lately.

Maybe something that can be extended in a form of an agent, similar to what has been done for `RetryAgent`: https://github.com/nodejs/undici/blob/ef25209858817e697fd7607850413357c2b3b7e2/lib/dispatcher/retry-agent.js#L6
